### PR TITLE
Remove _app.js from with-apollo-auth example

### PR DESCRIPTION
**Pull request by HaNdTriX**
## TLDR

This PR removes `_app.js` from `examples/with-apollo-auth`.
This allows automatic static optimization for pages that don't use apollo.

## Changelog

- Removed `_app.js` usage if favor of PageHoc.
- Migrate Class to Function Component
- Moved apollo init code into one file (`lib/apollo.js`)
- Remove `@apollo/react-ssr` from the client bundle
- Migrate all class Components to functional Components.
- Add improved display names
- Remove display name from production bundle
- Remove global fetch polyfill in favor of scoped ponyfill
- Add some more doc comments
- Use `HttpLink` from `apollo-boost`
- Simplify apollo links
- Moved proxy config next to apollos `HttpLink`
- Use esm import instead of require

Sorry for all the changes. But creating a PR for each one would have overcomplicated things.

## Related

#8504, #8508, #8538

### CODE REVIEW

**Comment by anarerdene** [comment]
So it means, we don't need _app.js when we using apollo ?

**Comment by HaNdTriX** [comment]
@anarerdene yes! Under the hood `withApollo` will run `getDataFromTree` on the new `AppTree`.

Please note that the new `withApollo`  HOC won't work with `App.js` anymore.

**Review by Timer — CHANGES_REQUESTED** [review]
Hey! Let's please use `import()` instead of `require()`, since mixing ESM and CJS is a discouraged behavior.

Thanks for taking this PR on!

**Comment by HaNdTriX** [comment]
@Timer good catch! Thanks!

File: examples/with-apollo-auth/lib/apollo.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+/**
+ * Always creates a new apollo client on the server
+ * Creates or reuses apollo client in the browser.
+ */
+function initApolloClient (...args) {
+  // Make sure to create a new client for every server-side request so that data
+  // isn't shared between connections (which would be bad)
+  if (typeof window === 'undefined') {
+    return createApolloClient(...args)
```
Comment (devlifter91) [inline examples/with-apollo-auth/lib/apollo.js]: Any way to make Typescript accept the `...args` ? It seems to get cranky with the spread operator...

**Review by lfades — APPROVED** [review]
@HaNdTriX Thank you for all the effort you've been putting into the examples 🚀  💯

File: examples/with-apollo-auth/lib/apollo.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  }
+
+  if (process.env.NODE_ENV !== 'production') {
+    // Find correct display name
+    const displayName =
+      PageComponent.displayName || PageComponent.name || 'Component'
+
+    // Warn if old way of installing apollo is used
+    if (displayName === 'App') {
+      console.warn('This withApollo HOC only works with PageComponents.')
```
Comment (raduchiriac) [inline examples/with-apollo-auth/lib/apollo.js]: Thank you @HaNdTriX for all your work.
 
What will happen if I still keep my `_app.js` and use
```
export default withApollo(MyApp);
```

Comment (HaNdTriX) [reply examples/with-apollo-auth/lib/apollo.js]: This won't work since the types of the params of `getInitialProps` and `appProps` differs in both use cases.

`App.getInitialProps<appContext > !== Page.getInitialProps<pageContext>`
`App<appProps> !== Page<pageProps>`

Please note that it would be possible to change the HOC to work with the App.
Nevertheless, we don't recommend this pattern since you will disable automatic static optimization for every route in your Next.js App.

Comment (raduchiriac) [reply examples/with-apollo-auth/lib/apollo.js]: Thank you it's a lot clearer now.
It's sad because all of my pages use Apollo. So it's cumbersome to add that everywhere now

Comment (HaNdTriX) [reply examples/with-apollo-auth/lib/apollo.js]: You can still use the old implementation.

Comment (iamandyk) [reply examples/with-apollo-auth/lib/apollo.js]: @HaNdTriX maybe we should add another example using `_app.js`? Most "apps" that don't have a lot of static pages will most likely find using `_app.js` easier with less boilerplate.

**Comment by raduchiriac** [comment]
This example has been removed but I still use it because it always worked for me.

@HaNdTriX is there a way to return `res` (and where)?
I would like to be able to do a `res.statusCode = 404` inside `getInitialProps()`.

**Comment by HaNdTriX** [comment]
Try the following code:

```js
const YourPageComponent = () => (
  <div>
    ...
  </div>
)


YourPageComponent.getInitialProps = (ctx) => {
  if (ctx.res) {
    ...
  }
}

export default withApollo(YourPageComponent)
```
